python/pylibraft/pylibraft/neighbors/eps_neighborhood.pyx

```diff
@@ -0,0 +1,341 @@
+
+    vd: array interface compliant object containing row sums of adj
+        shape (n_queries + 1). vd[n_queries] contains the total sum
+
+    Examples
+    --------
+    >>> import cupy as cp
+    >>> from pylibraft.common import DeviceResources
+    >>> from pylibraft.neighbors.eps_neighborhood import eps_neighbors
+    >>> handle = DeviceResources()
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> n_queries = 1000
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> queries = cp.random.random_sample((n_queries, n_features),
+    ...                                   dtype=cp.float32)
+    >>> eps = 0.1
+    >>> adj, vd = eps_neighbors(dataset, queries, eps, handle=handle)
+    >>> adj = cp.asarray(adj)
+    >>> vd = cp.asarray(vd)
+    >>> # pylibraft functions are often asynchronous so the
+    >>> # handle needs to be explicitly synchronized
+    >>> handle.sync()
+    """
+
+    if handle is None:
+        handle = DeviceResources()
+
+    dataset_cai = cai_wrapper(dataset)
+    queries_cai = cai_wrapper(queries)
+
+    # we require c-contiguous (rowmajor) inputs here
+    _check_input_array(dataset_cai, [np.dtype("float32")])
+    _check_input_array(queries_cai, [np.dtype("float32")],
+                       exp_cols=dataset_cai.shape[1])
+
+    n_queries = queries_cai.shape[0]
+    n_samples = dataset_cai.shape[0]
+
+    adj = device_ndarray.empty((n_queries, n_samples), dtype='bool')
+    vd = device_ndarray.empty((n_queries + 1, ), dtype='int64')
+    adj_cai = cai_wrapper(adj)
+    vd_cai = cai_wrapper(vd)
+
+    cdef device_resources* handle_ = \
+        <device_resources*><size_t>handle.getHandle()
+
+    vd_vector_view = make_device_vector_view(
+        <int64_t *><uintptr_t>vd_cai.data, <int64_t>vd_cai.shape[0])
+
+    if dataset_cai.dtype == np.float32:
+        with cuda_interruptible():
+            if method == "ball_cover":
+                c_eps_neighbors_l2_rbc(
+                    deref(handle_),
+                    get_dmv_float(dataset_cai, check_shape=True),
+                    get_dmv_float(queries_cai, check_shape=True),
+                    get_dmv_bool(adj_cai, check_shape=True),
+                    vd_vector_view,
+                    eps)
+            elif method == "brute":
+                c_eps_neighbors_l2(
+                    deref(handle_),
+                    get_dmv_float(dataset_cai, check_shape=True),
+                    get_dmv_float(queries_cai, check_shape=True),
+                    get_dmv_bool(adj_cai, check_shape=True),
+                    vd_vector_view,
+                    eps)
+            else:
+                raise ValueError("Unsupported method %s" % method)
+    else:
+        raise TypeError("dtype %s not supported" % dataset_cai.dtype)
+
+    return (adj, vd)
+
+
+@auto_sync_handle
+@auto_convert_output
+def eps_neighbors_sparse(RbcIndex rbc_index, queries, eps, handle=None):
```

Comment:
I'm a little conflicted about the layout in this namespace. I think it might be better and more consistent with the other `pylibraft` algorithms to separate the APIs by their associated index type. So, for example, `pylibraft.neighbors.rbc.eps_neighbors` and `pylibraft.neighbors.brute_force.eps_neighbors`. This is because, similar to knn, there are mutliple different algorithms for computing it and we separate each one of those different algorithms. `eps-nn` is similar in this way.

Reply:
> Splitting by namespace looks cleaner, but that would prevent us from making algorithm-decisions based on input shape or similar. This would leave the decision to the consumer (e.g. cuml). If that is feasible I can change the API accordingly.

@mfoerste4, I would be inclined to agree with you on this one if the two APIs on the RAFT side were able to be consolidated into a single function. Having different functions with different names which accept different arguments is probably not the best user experience. I do agree that it would enable that type of user experience in cuml, but we also want to make sure RAFT's APIs are clean and easy to use. 

The other detail here is that pylibraft *should* be a lightweight Python layer around RAFT's C++ functions, and with that being the case, we probably want to follow the design of the c++ side for consistency (note that I'm not referring to the versoin in raft::distance here. I would agree that should eventually be moved to the brute-force namespace).